### PR TITLE
bison: introduce compound assignment operators

### PR DESCRIPTION
**Pull request by mmarchini**

### CODE REVIEW

**Comment by dalehamel** [comment]
@mmarchini now that #452 is merged, I can revert my work from #439 along and replace it with an implementation along this pattern, unless you would prefer to.

Kudos on the elegant implementation :clap:

**Comment by mmarchini** [comment]
Feel free to do it :)

**Comment by dalehamel** [comment]
I put up #468 which dramatically simplifies the code while maintaining the functionality.

I'd be happy to finish this PR off for you if you have more important stuff on your list, as it should be basically the same work (update parser, write parser tests)

**Comment by mmarchini** [comment]
I probably won't be able to work on this before April, so go for it :)

**Comment by mmarchini** [comment]
@ajor @dalehamel this is ready for review now :)

File: src/parser.yy [hunk 1]
```diff
@@ -109,6 +121,7 @@ void yyerror(bpftrace::Driver &driver, const char *s);
 %type <std::string> ident
 
 %right ASSIGN
+%right PLUSASSIGN
```
Comment (ajor) [inline src/parser.yy]: Why does only this one need to be added here?

File: src/parser.yy [hunk 2]
```diff
@@ -60,6 +60,18 @@ void yyerror(bpftrace::Driver &driver, const char *s);
   LOR        "||"
   PLUS       "+"
   PLUSPLUS   "++"
+
+  LANDASSIGN  "&&="
```
Comment (ajor) [inline src/parser.yy]: Are there any other languages that have `&&=` and `||=`?

File: tests/semantic_analyser.cpp [hunk 3]
```diff
@@ -153,6 +153,76 @@ TEST(semantic_analyser, mismatched_call_types)
...
+  test("kprobe:f { $a ||= 0 }", 1);
+  test("kprobe:f { $a = 0; $a ||= 1 }", 0);
+  test("kprobe:f { @a ||= 1 }", 0);
+}
+
+TEST(semantic_analyser, compound_plus)
+{
+  test("kprobe:f { $a += 0 }", 1);
+  test("kprobe:f { $a = 0; $a += 1 }", 0);
+  test("kprobe:f { @a += 1 }", 0);
```
Comment (ajor) [inline tests/semantic_analyser.cpp]: Why is `+= 1` ok but `+= 0` is not?

File: src/parser.yy (hunk 1 above)
Comment (mmarchini) [reply src/parser.yy]: Don't remember, it's been a few months since I made this change. I'll investigate.

File: src/parser.yy (hunk 2 above)
Comment (mmarchini) [reply src/parser.yy]: Good question.

C/C++ doesn't have it, apparently: https://www.tutorialspoint.com/Compound-Assignment-Operators-in-Cplusplus

Neither does Python: https://www.programiz.com/python-programming/operators#assignment_operators

Or Java: https://www.geeksforgeeks.org/compound-assignment-operators-java/

Or JavaScript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators

They all have `<<=` and `>>=`, so I guess it makes more sense remove `&&=`/`||=` and add compound shift operators.

File: src/parser.yy (hunk 1 above)
Comment (mmarchini) [reply src/parser.yy]: Guess it wasn't needed after all.

**Comment by mmarchini** [comment]
> Could we get parser tests for each of these new operators too please?

Done

File: src/parser.yy (hunk 2 above)
Comment (danobi) [reply src/parser.yy]: Side note: funny, I was writing some C++ yesterday and I was hoping &&= existed.

**Comment by danobi** [comment]
Very nice; extremely clean.

**Comment by danobi** [comment]
Since this is a language change, would like to get @ajor 's or someone else's stamp too.

**Pull request merged**